eclair-core/src/main/scala/fr/acinq/eclair/io/PeerConnection.scala

```diff
@@ -0,0 +1,519 @@
+ * the [[Peer]] or a [[Client]] depending on whether it is an incoming or outgoing connection.
+ *
+ * There is one [[PeerConnection]] per TCP connection, and the [[PeerConnection]] actor dies when the underlying TCP connection
+ * dies.
+ *
+ * The [[PeerConnection]] will be responsible for handling:
+ * - the crypto handshake (BOLT 8)
+ * - the application handshake (exchange of [[Init]] messages, BOLT 1)
+ * - the routing messages (syncing and gossip, BOLT 7)
+ *
+ * Higher level messages (e.g. BOLT 2) are forwarded to the [[Peer]] actor, which is unique per logical peer (i.e. unique per
+ * node id).
+ *
+ * Most of the time there will only be one [[PeerConnection]] per logical peer, but there is no strict constraint and the
+ * [[Peer]] is responsible for deciding which [[PeerConnection]] stays and which should be killed.
+ *
+ * Created by PM on 11/03/2020.
+ */
+class PeerConnection(nodeParams: NodeParams, switchboard: ActorRef, router: ActorRef) extends FSMDiagnosticActorLogging[PeerConnection.State, PeerConnection.Data] {
+
+  import PeerConnection._
+
+  startWith(BEFORE_AUTH, Nothing)
+
+  when(BEFORE_AUTH) {
+    case Event(p: PendingAuth, _) =>
+      val transport = p.transport_opt match {
+        case Some(transport) => transport // used in tests to bypass encryption
+        case None =>
+          Metrics.PeerConnections.withTag(Tags.ConnectionState, Tags.ConnectionStates.Authenticating).increment()
+          context.actorOf(TransportHandler.props(
+            keyPair = KeyPair(nodeParams.nodeId.value, nodeParams.privateKey.value),
+            rs = p.remoteNodeId_opt.map(_.value),
+            connection = p.connection,
+            codec = LightningMessageCodecs.meteredLightningMessageCodec),
+            name = "transport")
+      }
+      context.watch(transport)
+      setTimer(AUTH_TIMER, AuthTimeout, nodeParams.authTimeout)
+      goto(AUTHENTICATING) using AuthenticatingData(p, transport)
+  }
+
+  when(AUTHENTICATING) {
+    case Event(TransportHandler.HandshakeCompleted(remoteNodeId), d: AuthenticatingData) =>
+      cancelTimer(AUTH_TIMER)
+      import d.pendingAuth.address
+      log.info(s"connection authenticated with $remoteNodeId@${address.getHostString}:${address.getPort} direction=${if (d.pendingAuth.outgoing) "outgoing" else "incoming"}")
+      Metrics.PeerConnections.withTag(Tags.ConnectionState, Tags.ConnectionStates.Authenticated).increment()
+      switchboard ! Authenticated(self, remoteNodeId, address, d.pendingAuth.outgoing, d.pendingAuth.origin_opt)
+      goto(BEFORE_INIT) using BeforeInitData(remoteNodeId, d.pendingAuth, d.transport)
+
+    case Event(AuthTimeout, _) =>
+      log.warning(s"authentication timed out after ${nodeParams.authTimeout}")
+      stop(FSM.Normal)
+  }
+
+  when(BEFORE_INIT) {
+    case Event(InitializeConnection(peer), d: BeforeInitData) =>
+      d.transport ! TransportHandler.Listener(self)
```

Comment:
Here we're relying on `switchboard` replying back with an `InitializeConnection` message, shouldn't we use a timeout for this state ?

Reply:
I've added state timeout only for adversarial scenarii. In this case if because of an internal problem we get stuck, then we can let our peer deal with that right? Otherwise we would want to be defensive we need a *lot* of timeouts all over the codebase?

Reply:
True, but having to split the connection workflow here (because for incoming connections switchboard does not have a peer at this point) does not feel 100% right